src/lib/Siteaccess/SiteaccessResolver.php

```diff
@@ -51,27 +52,53 @@ public function getSiteaccessesForLocation(
         Location $location,
         int $versionNo = null,
         string $languageCode = null
+    ): array {
+        return array_column(
+            $this->getSiteAccessesListForLocation($location, $versionNo, $languageCode),
+            'name'
+        );
+    }
+
+    /**
+     * @return \eZ\Publish\Core\MVC\Symfony\SiteAccess[]
+     */
+    public function getSiteAccessesListForLocation(
+        Location $location,
+        int $versionNo = null,
+        string $languageCode = null
     ): array {
         $contentInfo = $location->getContentInfo();
         $versionInfo = $this->contentService->loadVersionInfo($contentInfo, $versionNo);
         $languageCode = $languageCode ?? $contentInfo->mainLanguageCode;
 
-        $eligibleSiteaccesses = [];
-        foreach ($this->getSiteaccesses() as $siteaccess) {
-            $context = new SiteaccessPreviewVoterContext($location, $versionInfo, $siteaccess, $languageCode);
-            foreach ($this->siteaccessPreviewVoters as $siteaccessPreviewVoter) {
-                if ($siteaccessPreviewVoter->vote($context)) {
-                    $eligibleSiteaccesses[] = $siteaccess;
+        $eligibleSiteAccesses = [];
+        /** @var \eZ\Publish\Core\MVC\Symfony\SiteAccess $siteAccess */
+        foreach ($this->siteAccessService->getAll() as $siteAccess) {
+            $context = new SiteaccessPreviewVoterContext($location, $versionInfo, $siteAccess->name, $languageCode);
+            foreach ($this->siteAccessPreviewVoters as $siteAccessPreviewVoter) {
+                if ($siteAccessPreviewVoter->vote($context)) {
+                    $eligibleSiteAccesses[] = $siteAccess;
                     break;
                 }
             }
         }
 
-        return $eligibleSiteaccesses;
+        return $eligibleSiteAccesses;
     }
 
     public function getSiteaccesses(): array
     {
-        return $this->siteAccesses;
+        return $this->getSiteAccessList();
+    }
+
+    /**
+     * @return string[]
+     */
+    private function getSiteAccessList(): array
```

Comment:
Please add SA list as an argument and reuse it in `getSiteaccesses` and `getSiteaccessesForLocation` (we will avoid `array_column($saList, 'name')` duplication)